### ISSUE TITLE
readProcessStdout_ hangs after ~65 KB of output (only when compiled)

### ISSUE THREAD

**Issue by chrismwendt** [fpco/typed-process#11]
To repro:

- Clone https://github.com/chrismwendt/bug-in-typed-process
- Run `stack ghci` and call `main`, the process should complete
- Run `stack build ; stack exec app`, the process hangs

![2018-07-01 15 29 50](https://user-images.githubusercontent.com/1387653/42139485-af07f36e-7d43-11e8-8cd9-42011d332b26.gif)

I was also able to repro on lts-6.35 (typed-process-0.1.0.0) with `readProcess_`.

The number (~65 KB) is suspiciously close to 2^16, which points to some kind of buffering problem.

Someone told me that this could be caused by problems with bytestring inlining. Maybe compiling the program triggers these?

**Comment by snoyberg** [comment]
Can you try adding `ghc-options: -threaded` to the executable stanza?

**Comment by chrismwendt** [comment]
Huh, that fixes the issue :sparkles: But why? 🤔

**Comment by snoyberg** [comment]
Without the multithreaded runtime, the child process can fill up the buffer
and block on writing to it, while the parent process blocks on the child
process closing. With the multithreaded runtime, the parent process can
drain the buffer while another thread is blocked waiting.

**Comment by chrismwendt** [comment]
Is this because `process` uses the FFI, which blocks all other threads?

Ohhh I completely missed the warning at the top of [the documentation](https://hackage.haskell.org/package/typed-process-0.2.2.0/docs/System-Process-Typed.html):

![image](https://user-images.githubusercontent.com/1387653/42205268-cfdf509e-7e58-11e8-9472-88aa928dd41f.png)

Considering how badly this breaks, would it be appropriate for `process` or `typed-process` to emit a warning/error saying you should use [`-threaded`](https://downloads.haskell.org/~ghc/latest/docs/html/users_guide/phases.html#ghc-flag--threaded) when it detects that it's running on the single-threaded GHC runtime? The value of [`getNumProcessors`](http://hackage.haskell.org/package/base-4.11.1.0/docs/GHC-Conc.html#v:getNumProcessors) could be checked to see if it's greater than 1.

**Comment by nh2** [comment]
And there is another thing going on:
I was wondering why the problems of `#11` shouldn't be fixed automatically by the timer signal.
In my world, the timer signal should interrupt the blocking `wait4()` syscall, so that other Haskell threads can run (e.g. to consume the output of `sh` in the failing example).
But that doesn't happen.
Turns out in `man 7 signal` (on my Ubuntu 16.04) there is this feature that I didn't even know about:

```

   Interruption of system calls and library functions by signal handlers
       If a signal handler is invoked while a system call or library
       function call is blocked, then either:

       * the call is automatically restarted after the signal handler
         returns; or

       * the call fails with the error EINTR.

       Which of these two behaviors occurs depends on the interface and
       whether or not the signal handler was established using the
       SA_RESTART flag (see sigaction(2)).  The details vary across UNIX
       systems; below, the details for Linux.

       If a blocked call to one of the following interfaces is interrupted
       by a signal handler, then the call is automatically restarted after
       the signal handler returns if the SA_RESTART flag was used; otherwise
       the call fails with the error EINTR:

       * read(2), readv(2), write(2), writev(2), and ioctl(2) calls on
         "slow" devices.  A "slow" device is one where the I/O call may
         block for an indefinite time, for example, a terminal, pipe, or
         socket.  If an I/O call on a slow device has already transferred
         some data by the time it is interrupted by a signal handler, then
         the call will return a success status (normally, the number of
         bytes transferred).  Note that a (local) disk is not a slow device
         according to this definition; I/O operations on disk devices are
         not interrupted by signals.

       * open(2), if it can block (e.g., when opening a FIFO; see fifo(7)).

       * wait(2), wait3(2), wait4(2), waitid(2), and waitpid(2).

       * Socket interfaces: accept(2), connect(2), recv(2), recvfrom(2),
         recvmmsg(2), recvmsg(2), send(2), sendto(2), and sendmsg(2), unless
         a timeout has been set on the socket (see below).

       * File locking interfaces: flock(2) and the F_SETLKW and F_OFD_SETLKW
         operations of fcntl(2)

       * POSIX message queue interfaces: mq_receive(3), mq_timedreceive(3),
         mq_send(3), and mq_timedsend(3).

       * futex(2) FUTEX_WAIT (since Linux 2.6.22; beforehand, always failed
         with EINTR).

       * getrandom(2).

       * pthread_mutex_lock(3), pthread_cond_wait(3), and related APIs.

       * futex(2) FUTEX_WAIT_BITSET.

       * POSIX semaphore interfaces: sem_wait(3) and sem_timedwait(3) (since
         Linux 2.6.22; beforehand, always failed with EINTR).

       * read(2) from an inotify(7) file descriptor (since Linux 3.8;
         beforehand, always failed with EINTR).

       The following interfaces are never restarted after being interrupted
       by a signal handler, regardless of the use of SA_RESTART; they always
       fail with the error EINTR when interrupted by a signal handler:

       * "Input" socket interfaces, when a timeout (SO_RCVTIMEO) has been
         set on the socket using setsockopt(2): accept(2), recv(2),
         recvfrom(2), recvmmsg(2) (also with a non-NULL timeout argument),
         and recvmsg(2).

       * "Output" socket interfaces, when a timeout (SO_RCVTIMEO) has been
         set on the socket using setsockopt(2): connect(2), send(2),
         sendto(2), and sendmsg(2).

       * Interfaces used to wait for signals: pause(2), sigsuspend(2),
         sigtimedwait(2), and sigwaitinfo(2).

       * File descriptor multiplexing interfaces: epoll_wait(2),
         epoll_pwait(2), poll(2), ppoll(2), select(2), and pselect(2).

       * System V IPC interfaces: msgrcv(2), msgsnd(2), semop(2), and
         semtimedop(2).

       * Sleep interfaces: clock_nanosleep(2), nanosleep(2), and usleep(3).

       * io_getevents(2).

       The sleep(3) function is also never restarted if interrupted by a
       handler, but gives a success return: the number of seconds remaining
       to sleep.
```

So you can set `SA_RESTART` and ~libc~ the kernel will auto-loop around `EINTR` for you so you don't have to, as a convenience.
But that concept being in effect here means that the following lines in `process` don't do what they should:
https://github.com/haskell/process/blob/e250d7c579ae/System/Process.hs#L637
https://github.com/haskell/process/blob/e250d7c579ae/System/Process.hs#L779
https://github.com/haskell/process/blob/e250d7c579ae/cbits/runProcess.c#L438-L441
Somebody carefully used `foreign import ccall interruptible` and ensured than `EINTR` would be returned up to Haskell so that `waitpid()` can be interrupted and Haskell can run in that case.
But this can never happen, because with `SA_RESTART` being set (which it seems to be in my case), `waitpid()` won't return at all when the `SIGVTALRM` comes in.

---

OK, I understand the logic behind that now.
The above logic (the `SA_RESTART` flag) can be set independently for each signal.
GHC uses `SIGVTALRM` for the timer signal and `SIGPIPE` for implementing `interruptible`.
GHC sets `SA_RESTART` only for `SIGVTALRM`, probably with the idea that C functions should be as little as possible affected by the timer signal.
It does it here: https://github.com/ghc/ghc/blob/51abb1c88b/rts/posix/Signals.c#L644-L652
So `interruptible` will work to interrupt `waitpit()`, but the timer signal can't interrupt it.

---

`SA_RESTART` was initially introduced in GHC here: https://github.com/ghc/ghc/commit/1494f895d69a33df2f8d855f73943faec5c6858e to work around some problems with the `readline` library for the `-threaded` RTS.

---

Also interesting: [Here](https://stackoverflow.com/questions/43487950/how-to-handle-signals-when-process-is-waiting-in-waitpid/43540451#43540451) a user claims that

> My Linux flavor(RedHat-like distro, 2.6.39 kernel) doesn't obey SA_RESTART for `waitpid` system call.

So it looks like you can't rely on `SA_RESTART` working. But that may be OK because `SA_RESTART` for `SIGVTALRM` really looks like a performance optimisation that just saves us some context switches when it works.

**Comment by snoyberg** [comment]
Fixed by #12
